### ISSUE TITLE
Tests get false positive if you yield without wrapping (in coffeescript)

### ISSUE THREAD

**Issue by phoenixeliot** [jasmine/jasmine#1233]
Version: jasmine-core@2.4.1
Environment: node, coffeescript
Run via `npm run jasmine`

This is a problem especially with Coffeescript, which will "helpfully" automatically convert functions to generators for you if you use `yield` in them.

We normally use this slight modification of `co.wrap` around all our `it` test functions, so that we can `yield` inside them.

```
co = require 'co'

module.exports = utils =
  wrap: (gen) ->
    fn = co.wrap(gen)
    return (done) ->
      fn.apply(@, [done]).catch (err) -> done.fail(err)
```

The problem is that if we accidentally leave off `utils.wrap`, the test still passes silently, but none of the expectations were actually run. This seems to happen any time the callback that `it` gets is a generator function.

Here's a spec that works correctly (by failing)
```
it "doesn't silently pass generator functions", utils.wrap (done) ->
  a = yield Promise.resolve("a")
  expect(false).toBe(true)
  done()
```

Here's a spec that demonstrates the problem (by passing)
```
it 'silently passes generator functions', ->
  a = yield Promise.resolve("a")
  expect(false).toBe(true)
```

The problem doesn't occur if you use `(done) ->`, (it correctly fails due to timeout), but because `yield`ing tests look synchronous, we failed to notice that in several cases.

This is what we've put in place for now to prevent this from ever happening, in our spec helper.js file, which successfully revealed three old tests which were not actually being run due to this problem. It would be great if something like this were put into Jasmine proper.
```
var oldIt = global.it;
global.it = function(description, testFn) {
  if(testFn.constructor.name === 'GeneratorFunction'){
    throw(Error("You didn't wrap a generator function! Do it like this: it 'does a thing', utils.wrap (done) ->"))
  }
  oldIt.apply(jasmine.getEnv(), arguments)
}
```

**Comment by slackersoft** [comment]
It looks like when you fail to wrap your spec functions, that the function passed into Jasmine doesn't have any arguments so Jasmine has no way to tell that it is supposed to be async. This means that as soon as you `yield` Jasmine moves on to the next spec in your suite.

Without a large refactoring to support Promises as return types or something (like #965) which we're not sure we want in Jasmine-Core at the moment, I'm not sure how we would support generator functions.

**Comment by phoenixeliot** [comment]
(whoops, dropped this from my radar forever)

The thing we ended up doing was something like this, which seems like it could be dropped into Jasmine fairly easily without implementing full support of unwrapped generators. So at least now we get a test failing instead of being ignored if we make that mistake.

```
var oldIt = global.it;
global.it = function(description, testFn) {
  if(testFn.constructor.name === 'GeneratorFunction'){
    throw(Error("Function passed to 'it' must not be a generator function"))
  }
  oldIt.apply(jasmine.getEnv(), arguments)
}
```

**Comment by slackersoft** [comment]
I think this actually gets fixed with #1222 which checks that `it`, `describe`, etc. all receive a `function` where they expect it. This hasn't been released yet, but you should be able to run off of master to check if this solves your problem.

Closing this. Thanks for using Jasmine!
